warehouse/query-core/src/main/java/datawave/query/index/lookup/IndexInfo.java

```diff
@@ -143,75 +137,44 @@ public IndexInfo union(IndexInfo o) {
         
         if (null != first.myNode && first.myNode != o.myNode) {
             
             JexlNode sourceNode = getSourceNode(first.myNode);
             JexlNode topLevelOr = getOrNode(sourceNode);
             
             if (null == topLevelOr) {
-                topLevelOr = sourceNode;
-                // add the source node
-                nodesMap.put(nodeToKey(sourceNode), first.myNode);
+                nodeSet.add(sourceNode);
             } else {
                 for (int i = 0; i < topLevelOr.jjtGetNumChildren(); i++) {
-                    JexlNode baseNode = getSourceNode(topLevelOr.jjtGetChild(i));
-                    nodesMap.put(nodeToKey(baseNode), topLevelOr.jjtGetChild(i));
+                    nodeSet.add(topLevelOr.jjtGetChild(i));
                 }
             }
         }
         
         if (null != o.myNode) {
-            
             JexlNode sourceNode = getSourceNode(o.myNode);
             JexlNode topLevelOr = getOrNode(sourceNode);
             
             if (null == topLevelOr) {
-                topLevelOr = sourceNode;
-                // add the source node
-                nodesMap.put(nodeToKey(topLevelOr), o.myNode);
+                nodeSet.add(o.myNode);
             } else {
                 for (int i = 0; i < topLevelOr.jjtGetNumChildren(); i++) {
-                    JexlNode baseNode = getSourceNode(topLevelOr.jjtGetChild(i));
-                    nodesMap.put(nodeToKey(baseNode), topLevelOr.jjtGetChild(i));
-                }
-            }
-        }
-        
-        for (JexlNode node : delayedNodes) {
-            JexlNode baseNode = getSourceNode(node);
-            nodesMap.put(nodeToKey(baseNode), node);
-        }
-        
-        for (String key : nodesMap.keySet()) {
-            Collection<JexlNode> nodeColl = nodesMap.get(key);
-            JexlNode delayedNode = null;
-            if (nodeColl.size() > 1) {
-                for (JexlNode node : nodeColl) {
-                    if (isDelayed(node)) {
-                        delayedNode = node;
-                        break;
-                    }
+                    nodeSet.add(topLevelOr.jjtGetChild(i));
                 }
             }
-            if (null != delayedNode)
-                internalNodeList.add(delayedNode);
-            else
-                internalNodeList.add(nodeColl.iterator().next());
         }
         
-        // ensure that we already aren't in there
+        // Add delayed nodes.
+        nodeSet.addAll(delayedNodes);
         
         merged.count = -1;
         merged.uids = ImmutableSortedSet.of();
-        if (internalNodeList.isEmpty()) {
-            
+        if (nodeSet.isEmpty()) {
             merged.myNode = null;
         } else {
-            
-            merged.myNode = JexlNodeFactory.createUnwrappedOrNode(internalNodeList);
+            merged.myNode = JexlNodeFactory.createUnwrappedOrNode(nodeSet.getNodes());
```

Comment:
The original logic was only sticking only putting a delayed node or one of the other nodes into merge.myNode.  What is the effect of including all of the nodes here?

Reply:
The original logic is still preserved, although it has been pushed down into the JexlNodeSet. Node keys for delayed nodes were (and still are) built from source nodes. If the JexlNodeSet encounters a delayed version of a source node, it will retain the delayed node over the source node.

The effect of calling nodeSet.getNodes() is the same as creating the unwrapped ORNode from the internal node list.

Reply:
To me, it looks like the original logic checked the list of nodes stored for a given key to see if any of them were delayed.  If a delayed version was found, that version would be added to the internalNodeList.  If none of them were delayed, it would just add the first node for that key.  With the new logic, we are no longer dealing with a single value map, not a multimap.  Because of that you added logic which will ensure that if a marked node is found for a given key, that node will be the one saved in the map.  So, unless I'm missing something, this new implementation looks like it matches the previous implementation.